File: android/src/main/java/expo/modules/libsignal/LibsignalErrors.kt
package expo.modules.libsignal

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import org.signal.libsignal.protocol.DuplicateMessageException
import org.signal.libsignal.protocol.InvalidKeyException
import org.signal.libsignal.protocol.InvalidKeyIdException
import org.signal.libsignal.protocol.InvalidMessageException
import org.signal.libsignal.protocol.NoSessionException
import org.signal.libsignal.protocol.UntrustedIdentityException

class LibsignalErrorPayload : Record {
  @Field var kind: String = "Generic"
  @Field var message: String = ""
}

fun mapSignalError(error: Throwable): LibsignalErrorPayload {
  val payload = LibsignalErrorPayload()
  payload.message = error.message ?: error.javaClass.simpleName
  payload.kind = when (error) {
    is UntrustedIdentityException -> "UntrustedIdentity"
    is InvalidMessageException -> "InvalidMessage"
    is NoSessionException -> "SessionNotFound"
    is InvalidKeyException, is InvalidKeyIdException -> "InvalidKey"
    is DuplicateMessageException -> "DuplicateMessage"
    else -> "Generic"
  }
  return payload
}

File: ios/LibsignalErrors.swift
import Foundation
import ExpoModulesCore
import LibSignalClient

enum LibsignalErrorKind: String {
  case untrustedIdentity = "UntrustedIdentity"
  case invalidMessage = "InvalidMessage"
  case sessionNotFound = "SessionNotFound"
  case invalidKey = "InvalidKey"
  case duplicateMessage = "DuplicateMessage"
  case generic = "Generic"
}

struct LibsignalErrorPayload: Record {
  @Field var kind: String = "Generic"
  @Field var message: String = ""
}

func mapSignalError(_ error: Error) -> LibsignalErrorPayload {
  let payload = LibsignalErrorPayload()
  payload.message = "\(error)"

  if let signalError = error as? SignalError {
    switch signalError {
    case .untrustedIdentity:
      payload.kind = LibsignalErrorKind.untrustedIdentity.rawValue
    case .invalidMessage:
      payload.kind = LibsignalErrorKind.invalidMessage.rawValue
    case .sessionNotFound:
      payload.kind = LibsignalErrorKind.sessionNotFound.rawValue
    case .invalidKey, .invalidKeyIdentifier:
      payload.kind = LibsignalErrorKind.invalidKey.rawValue
    case .duplicatedMessage:
      payload.kind = LibsignalErrorKind.duplicateMessage.rawValue
    default:
      payload.kind = LibsignalErrorKind.generic.rawValue
    }
  } else {
    payload.kind = LibsignalErrorKind.generic.rawValue
  }
  return payload
}
